language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/Primitive.scala

```diff
@@ -184,29 +160,27 @@ private[client] object OnlyPrimitive extends Primitive {
   object TemplateId extends TemplateIdApi {
     // the ledger api still uses names with only dots in them, while QualifiedName.toString
     // separates the module and the name in the module with colon.
-    override def apply[Tpl <: Template[Tpl]](
+    def apply[Tpl <: Template[Tpl]](
         packageId: String,
         moduleName: String,
         entityName: String): TemplateId[Tpl] =
       ApiTypes.TemplateId(
-        rpcvalue.Identifier(
-          packageId = packageId,
-          name = s"$moduleName.$entityName",
-          moduleName = moduleName,
-          entityName = entityName))
+        rpcvalue
+          .Identifier(packageId = packageId, moduleName = moduleName, entityName = entityName))
 
     private[binding] override def substEx[F[_]](fa: F[rpcvalue.Identifier]) =
       ApiTypes.TemplateId subst fa
 
-    override def unapply[Tpl](t: TemplateId[Tpl]): Option[(String, String, String)] =
+    def unapply[Tpl](t: TemplateId[Tpl]): Option[(String, String, String)] =
       // TODO SC DEL-6727 use this instead with daml-lf value interface
       // rpcvalue.Identifier unapply t.unwrap
       t.unwrap match {
-        case rpcvalue.Identifier(packageId, _, moduleName, entityName)
+        case rpcvalue.Identifier(packageId, moduleName, entityName)
             if moduleName.nonEmpty && entityName.nonEmpty =>
           Some((packageId, moduleName, entityName))
         case _ => None
```

Comment:
```suggestion
```

Reply:
I see the reasoning behind your suggestion but I'd suggest you keep the default case and throw a telling exception: `MatchError`s tend to be very vague and it would be easier to understand the issue if the exception had a meaningful message based on the context.

Reply:
There is no possibility of a `MatchError` in the set of changes I've suggested.

Reply:
Now there isn't, which is what makes the situation exceptional. In general, I hardly see good reasons to leave generic `MatchError`s with very error messages leaving a lot to be said being possibly thrown at runtime when a more meaningful and contextualized. If there is not possibility that the match couldn't fail, wouldn't it be possible to simply do something like

```scala
val unwrapped = t.unwrap
Some((unwrapped.packageId, unwrapped.moduleName, unwrapped.entityName))
```

Seems more direct and readable.

Reply:
> Seems more direct and readable.

You are right but I can't express that in GH change suggestions :)

Reply:
```scala
 val rpcvalue.Identifier(packageId, moduleName, entityName) = t.unwrap
 Some((packageId, moduleName, entityName))
``